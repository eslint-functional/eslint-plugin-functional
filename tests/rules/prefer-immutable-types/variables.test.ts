import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/prefer-immutable-types";

import { typescriptConfig } from "../../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("reports non-immutable set variables", async () => {
      const invalidResult = await invalid({
        code: "const foo: ReadonlySet<string> = {} as any;",
        errors: ["variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
    });

    it("reports non-immutable map variables", async () => {
      const invalidResult = await invalid({
        code: "const foo: ReadonlyMap<string, string> = {} as any;",
        errors: ["variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
    });

    it("reports mutable records variables", async () => {
      const invalidResult = await invalid({
        code: "const foo: { foo: string } = {} as any;",
        errors: ["variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
    });

    it("reports mutable records variables and suggests a fix for ReadonlyShallow", async () => {
      const invalidResult = await invalid({
        code: "const foo: { foo: string } = {} as any;",
        options: [{ variables: "ReadonlyShallow" }],
        errors: ["variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("doesn't report type assertion variables", async () => {
      await valid({
        code: "function foo(arg0: { foo: string | number }, arg1: { foo: string | number }): arg0 is { foo: number } {}",
        options: [
          {
            parameters: false,
            variables: "ReadonlyShallow",
          },
        ],
      });
    });

    it("suggest multiple fixes for collections for ReadonlyShallow", async () => {
      const invalidResult = await invalid({
        code: dedent`
          const foo: Array<string> = {} as any;
          const foo: string[] = {} as any;
          const foo: Set<string> = {} as any;
          const foo: Map<string, string> = {} as any;
          const foo: ReadonlyArray<string> = {} as any;
          const foo: readonly string[] = {} as any;
          const foo: ReadonlySet<string> = {} as any;
          const foo: ReadonlyMap<string, string> = {} as any;
        `,
        options: [{ variables: "ReadonlyShallow" }],
        errors: ["variable", "variable", "variable", "variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(2);
      }
    });

    it("reports mutable class parameter properties", async () => {
      const invalidResult = await invalid({
        code: dedent`
          class Klass {
            readonly foo: { foo: number };
            private readonly bar: { foo: number };
            static readonly baz: { foo: number };
            private static readonly qux: { foo: number };
          }
        `,
        errors: ["propertyImmutability", "propertyImmutability", "propertyImmutability", "propertyImmutability"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
    });

    it("allows for user suggestions", async () => {
      const invalidResult = await invalid({
        code: "const foo: { foo: string } = {} as any;",
        options: [
          {
            variables: "ReadonlyDeep",
            suggestions: {
              ReadonlyDeep: [
                [
                  {
                    pattern: "^(?!ReadonlyDeep)(?:Readonly<(.+)>|(.+))$",
                    replace: "ReadonlyDeep<$1$2>",
                  },
                ],
              ],
            },
          },
        ],
        errors: ["variable"],
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("allows for user fixes", async () => {
      const invalidResult = await invalid({
        code: dedent`
          const foo: Array<string> = {} as any;
          const foo: string[] = {} as any;
          const foo: Set<string> = {} as any;
          const foo: Map<string, string> = {} as any;
          const foo: ReadonlyArray<string> = {} as any;
          const foo: readonly string[] = {} as any;
          const foo: ReadonlySet<string> = {} as any;
          const foo: ReadonlyMap<string, string> = {} as any;
        `,
        options: [
          {
            variables: "Immutable",
            fixer: {
              Immutable: [
                {
                  pattern: "^(?:Readonly)?(Set|Map)<(.+)>$",
                  replace: "Readonly<Readonly$1<$2>>",
                },
              ],
            },
          },
        ],
        errors: ["variable", "variable", "variable", "variable", "variable", "variable", "variable", "variable"],
        verifyAfterFix: false, // "fix" doesn't fix arrays so they will still report.
      });
      expect(invalidResult.result.messages).toMatchSnapshot();
    });

    it.each([[[{ variables: "ReadonlyShallow" }]], [[{ variables: "ReadonlyDeep" }]], [[{ variables: "Immutable" }]]])(
      "doesn't reports valid variables",
      async (options) => {
        await valid({
          code: "const foo: boolean = {} as any;",
          options,
        });
        await valid({
          code: "const foo: true = {} as any;",
          options,
        });
        await valid({
          code: "const foo: string = {} as any;",
          options,
        });
        await valid({
          code: "const foo: 'bar' = {} as any;",
          options,
        });
        await valid({
          code: "const foo: undefined = {} as any;",
          options,
        });
        await valid({
          code: "const foo: readonly string[] = {} as any;",
          options,
          settings: {
            immutability: {
              overrides: [
                {
                  type: { from: "lib", name: "ReadonlyArray" },
                  to: "Immutable",
                },
              ],
            },
          },
        });
        await valid({
          code: "const foo: { readonly foo: string } = {} as any;",
          options,
        });
        await valid({
          code: "const foo: { readonly foo: { readonly bar: number } } = {} as any;",
          options,
        });
        await valid({
          code: "const foo: Readonly<ReadonlySet<string>> = {} as any;",
          options,
        });
        await valid({
          code: "const foo: Readonly<ReadonlyMap<string, string>> = {} as any;",
          options,
        });
        await valid({
          code: "function foo(arg: { foo: string | number }): arg is { foo: number } {}",
          options,
        });
        await valid({
          code: "const [a, b] = [1, 2];",
          options,
        });
        await valid({
          code: "const { a, b } = { a: 1, b: 2 };",
          options,
        });

        if (options[0]!.variables !== "Immutable") {
          await valid({
            code: "const foo: { foo(): void } = {} as any;",
            options,
          });
          await valid({
            code: "const foo: ReadonlyArray<string> = {} as any;",
            options,
          });
          await valid({
            code: "const foo: readonly [string, number] = {} as any;",
            options,
          });
          await valid({
            code: "const foo: Readonly<[string, number]> = {} as any;",
            options,
          });
          await valid({
            code: "const foo: { foo: () => void } = {} as any;",
            options,
          });
          await valid({
            code: "const foo: ReadonlySet<string> = {} as any;",
            options,
          });
          await valid({
            code: "const foo: ReadonlyMap<string, string> = {} as any;",
            options,
          });
        }
      },
    );

    describe("options", () => {
      describe("ignoreClasses", () => {
        it("doesn't report class fields", async () => {
          await valid({
            code: dedent`
              class Klass {
                foo: number;
                private bar: number;
                static baz: number;
                private static qux: number;
              }
            `,
            options: [{ ignoreClasses: true }],
          });
        });
      });

      describe("ignoreInFunctions", () => {
        it("doesn't report variables in functions", async () => {
          await valid({
            code: dedent`
              function foo() {
                let foo: {
                  a: { foo: number },
                  b: string[],
                  c: { [key: string]: string[] },
                  [key: string]: any,
                }
              };
            `,
            options: [{ variables: { ignoreInFunctions: true } }],
          });
        });
      });

      describe("ignoreNamePattern", () => {
        it("doesn't report ignored names", async () => {
          await valid({
            code: "let mutableFoo: string[] = [];",
            options: [{ ignoreNamePattern: "^mutable" }],
          });
        });
      });

      describe("ignoreTypePattern", () => {
        it("doesn't report ignored types", async () => {
          await valid({
            code: dedent`
              let fooMutable: Readonly<
                string[]
              > = [];
            `,
            options: [{ ignoreTypePattern: "^Readonly<.+>$" }],
          });
        });
      });
    });
  });
});
