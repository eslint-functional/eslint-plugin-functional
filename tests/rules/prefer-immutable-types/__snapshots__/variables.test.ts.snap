// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-immutable-types > typescript > allows for user fixes 1`] = `
{
  "fixed": true,
  "messages": [
    {
      "column": 7,
      "endColumn": 25,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 20,
      "endLine": 2,
      "line": 2,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 23,
      "endLine": 3,
      "fix": {
        "range": [
          82,
          93,
        ],
        "text": "Readonly<ReadonlySet<string>>",
      },
      "line": 3,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 31,
      "endLine": 4,
      "fix": {
        "range": [
          118,
          137,
        ],
        "text": "Readonly<ReadonlyMap<string, string>>",
      },
      "line": 4,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 33,
      "endLine": 5,
      "line": 5,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 29,
      "endLine": 6,
      "line": 6,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 31,
      "endLine": 7,
      "fix": {
        "range": [
          250,
          269,
        ],
        "text": "Readonly<ReadonlySet<string>>",
      },
      "line": 7,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 7,
      "endColumn": 39,
      "endLine": 8,
      "fix": {
        "range": [
          294,
          321,
        ],
        "text": "Readonly<ReadonlyMap<string, string>>",
      },
      "line": 8,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "const foo: Array<string> = {} as any;
const foo: string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;
const foo: ReadonlyArray<string> = {} as any;
const foo: readonly string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;",
  "steps": [
    {
      "fixed": true,
      "messages": [
        {
          "column": 7,
          "endColumn": 25,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 20,
          "endLine": 2,
          "line": 2,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 33,
          "endLine": 5,
          "line": 5,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 29,
          "endLine": 6,
          "line": 6,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "const foo: Array<string> = {} as any;
const foo: string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;
const foo: ReadonlyArray<string> = {} as any;
const foo: readonly string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;",
    },
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 25,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 20,
          "endLine": 2,
          "line": 2,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 33,
          "endLine": 5,
          "line": 5,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 7,
          "endColumn": 29,
          "endLine": 6,
          "line": 6,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "const foo: Array<string> = {} as any;
const foo: string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;
const foo: ReadonlyArray<string> = {} as any;
const foo: readonly string[] = {} as any;
const foo: Readonly<ReadonlySet<string>> = {} as any;
const foo: Readonly<ReadonlyMap<string, string>> = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > allows for user suggestions 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 27,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "ReadonlyDeep" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Replace with: ReadonlyDeep<{ foo: string }>",
          },
          "desc": "Replace with: ReadonlyDeep<{ foo: string }>",
          "fix": {
            "range": [
              11,
              26,
            ],
            "text": "ReadonlyDeep<{ foo: string }>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "const foo: { foo: string } = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 27,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "ReadonlyDeep" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Replace with: ReadonlyDeep<{ foo: string }>",
              },
              "desc": "Replace with: ReadonlyDeep<{ foo: string }>",
              "fix": {
                "range": [
                  11,
                  26,
                ],
                "text": "ReadonlyDeep<{ foo: string }>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "const foo: { foo: string } = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports mutable class parameter properties 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 3,
      "endColumn": 33,
      "endLine": 2,
      "line": 2,
      "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "propertyImmutability",
      "nodeType": "PropertyDefinition",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 3,
      "endColumn": 41,
      "endLine": 3,
      "line": 3,
      "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "propertyImmutability",
      "nodeType": "PropertyDefinition",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 3,
      "endColumn": 40,
      "endLine": 4,
      "line": 4,
      "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "propertyImmutability",
      "nodeType": "PropertyDefinition",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 3,
      "endColumn": 48,
      "endLine": 5,
      "line": 5,
      "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "propertyImmutability",
      "nodeType": "PropertyDefinition",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "class Klass {
  readonly foo: { foo: number };
  private readonly bar: { foo: number };
  static readonly baz: { foo: number };
  private static readonly qux: { foo: number };
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 3,
          "endColumn": 33,
          "endLine": 2,
          "line": 2,
          "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "propertyImmutability",
          "nodeType": "PropertyDefinition",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 3,
          "endColumn": 41,
          "endLine": 3,
          "line": 3,
          "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "propertyImmutability",
          "nodeType": "PropertyDefinition",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 3,
          "endColumn": 40,
          "endLine": 4,
          "line": 4,
          "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "propertyImmutability",
          "nodeType": "PropertyDefinition",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 3,
          "endColumn": 48,
          "endLine": 5,
          "line": 5,
          "message": "Property should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "propertyImmutability",
          "nodeType": "PropertyDefinition",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "class Klass {
  readonly foo: { foo: number };
  private readonly bar: { foo: number };
  static readonly baz: { foo: number };
  private static readonly qux: { foo: number };
}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports mutable records variables 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 27,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "const foo: { foo: string } = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 27,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "const foo: { foo: string } = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports mutable records variables and suggests a fix for ReadonlyShallow 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 27,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              11,
              26,
            ],
            "text": "Readonly<{ foo: string }>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "const foo: { foo: string } = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 27,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  11,
                  26,
                ],
                "text": "Readonly<{ foo: string }>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "const foo: { foo: string } = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports non-immutable map variables 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 39,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "const foo: ReadonlyMap<string, string> = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 39,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "const foo: ReadonlyMap<string, string> = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports non-immutable set variables 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 31,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "const foo: ReadonlySet<string> = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 31,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "const foo: ReadonlySet<string> = {} as any;",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > suggest multiple fixes for collections for ReadonlyShallow 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 7,
      "endColumn": 25,
      "endLine": 1,
      "line": 1,
      "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlyArray instead of Array.",
          },
          "desc": "Use ReadonlyArray instead of Array.",
          "fix": {
            "range": [
              11,
              24,
            ],
            "text": "ReadonlyArray<string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              11,
              24,
            ],
            "text": "Readonly<Array<string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 7,
      "endColumn": 20,
      "endLine": 2,
      "line": 2,
      "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Prepend with readonly.",
          },
          "desc": "Prepend with readonly.",
          "fix": {
            "range": [
              49,
              57,
            ],
            "text": "readonly string[]",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              49,
              57,
            ],
            "text": "Readonly<string[]>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 7,
      "endColumn": 23,
      "endLine": 3,
      "line": 3,
      "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlySet instead of Set.",
          },
          "desc": "Use ReadonlySet instead of Set.",
          "fix": {
            "range": [
              82,
              93,
            ],
            "text": "ReadonlySet<string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              82,
              93,
            ],
            "text": "Readonly<Set<string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 7,
      "endColumn": 31,
      "endLine": 4,
      "line": 4,
      "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "variable",
      "nodeType": "Identifier",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlyMap instead of Map.",
          },
          "desc": "Use ReadonlyMap instead of Map.",
          "fix": {
            "range": [
              118,
              137,
            ],
            "text": "ReadonlyMap<string, string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              118,
              137,
            ],
            "text": "Readonly<Map<string, string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "const foo: Array<string> = {} as any;
const foo: string[] = {} as any;
const foo: Set<string> = {} as any;
const foo: Map<string, string> = {} as any;
const foo: ReadonlyArray<string> = {} as any;
const foo: readonly string[] = {} as any;
const foo: ReadonlySet<string> = {} as any;
const foo: ReadonlyMap<string, string> = {} as any;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 7,
          "endColumn": 25,
          "endLine": 1,
          "line": 1,
          "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlyArray instead of Array.",
              },
              "desc": "Use ReadonlyArray instead of Array.",
              "fix": {
                "range": [
                  11,
                  24,
                ],
                "text": "ReadonlyArray<string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  11,
                  24,
                ],
                "text": "Readonly<Array<string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 7,
          "endColumn": 20,
          "endLine": 2,
          "line": 2,
          "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Prepend with readonly.",
              },
              "desc": "Prepend with readonly.",
              "fix": {
                "range": [
                  49,
                  57,
                ],
                "text": "readonly string[]",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  49,
                  57,
                ],
                "text": "Readonly<string[]>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 7,
          "endColumn": 23,
          "endLine": 3,
          "line": 3,
          "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlySet instead of Set.",
              },
              "desc": "Use ReadonlySet instead of Set.",
              "fix": {
                "range": [
                  82,
                  93,
                ],
                "text": "ReadonlySet<string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  82,
                  93,
                ],
                "text": "Readonly<Set<string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 7,
          "endColumn": 31,
          "endLine": 4,
          "line": 4,
          "message": "Variable should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "variable",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlyMap instead of Map.",
              },
              "desc": "Use ReadonlyMap instead of Map.",
              "fix": {
                "range": [
                  118,
                  137,
                ],
                "text": "ReadonlyMap<string, string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  118,
                  137,
                ],
                "text": "Readonly<Map<string, string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "const foo: Array<string> = {} as any;
const foo: string[] = {} as any;
const foo: Set<string> = {} as any;
const foo: Map<string, string> = {} as any;
const foo: ReadonlyArray<string> = {} as any;
const foo: readonly string[] = {} as any;
const foo: ReadonlySet<string> = {} as any;
const foo: ReadonlyMap<string, string> = {} as any;",
    },
  ],
}
`;
